### PR TITLE
Add support for topk

### LINKED ISSUE #23: topk
Hello,

I notice that pytorch and tensorflow both have a `topk` function, which can be faster than a full sort if k is small. With numpy, it can be emulated using (arg)partition. I think it would be convenient if eagerpy provided a wrapper for this, I have 2 unrelated pieces of code where I would be likely to use it.

(of course it is disappointing that tensorflow does not seem to support the argument `dim`, but doing some transposition around the call is easy enough)

### PR DESCRIPTION
**Pull request by mglisse**
Fix #23.
This adds a minimal version of topk. In the future, it would be good to add some version of the `dim` and `largest` arguments from `torch.topk` (I use them), but I wanted to start with the basic functionality. The numpy version is a bit more complicated than one might hope, and jax complicates it further by not supporting `argpartition` and silently behaving differently in `take`.
I wasn't sure about the documentation, I searched for "sort" in docs/ and only found argsort in api/tensor.md, so I added topk there...

### CODE REVIEW

**Comment by jonasrauber** [comment]
Thanks, this is great! I will add some minor change requests, looking forward to merging it.

File: eagerpy/tensor/tensorflow.py [hunk 1]
```diff
@@ -176,6 +176,12 @@ def argsort(self: TensorType, axis: Optional[int] = -1) -> TensorType:
     def sort(self: TensorType, axis: Optional[int] = -1) -> TensorType:
         return type(self)(tf.sort(self.raw, axis=axis))
 
+    def topk(
+        self: TensorType, k: int, sorted: bool = True
+    ) -> Tuple[TensorType, TensorType]:
+        pair = tf.math.top_k(self.raw, k, sorted=sorted)
```
Comment (jonasrauber) [inline eagerpy/tensor/tensorflow.py:182]: You can use tuple unpacking instead: `values, indices = tf.math…`.
It's more descriptive and avoids tuple indexing in the next line.

File: eagerpy/tensor/pytorch.py [hunk 2]
```diff
@@ -203,6 +203,12 @@ def argsort(self: TensorType, axis: int = -1) -> TensorType:
     def sort(self: TensorType, axis: int = -1) -> TensorType:
         return type(self)(self.raw.sort(dim=axis).values)  # type: ignore
 
+    def topk(
+        self: TensorType, k: int, sorted: bool = True
+    ) -> Tuple[TensorType, TensorType]:
+        pair = self.raw.topk(k, sorted=sorted)
```
Comment (jonasrauber) [inline eagerpy/tensor/pytorch.py:209]: You can use tuple unpacking instead: `values, indices = self.raw…`.
It's more descriptive and avoids tuple indexing in the next line.

File: tests/test_main.py [hunk 3]
```diff
@@ -1206,6 +1206,18 @@ def test_sort(dummy: Tensor) -> Tensor:
     return ep.sort(t)
 
 
+@compare_all
+def test_topk_0(dummy: Tensor) -> Tensor:
```
Comment (jonasrauber) [inline tests/test_main.py:1210]: Could you rename the tests to end with `_values` and `_indices` instead of `0` and `1`.

File: tests/test_main.py [hunk 4]
```diff
@@ -1206,6 +1206,18 @@ def test_sort(dummy: Tensor) -> Tensor:
...
+    t = ep.arange(dummy, 27).float32().reshape((3, 3, 3))
```
Comment (jonasrauber) [inline tests/test_main.py:1211]: I think it would be good to test on something that's not already sorted.

File: tests/test_main.py [hunk 5]
```diff
@@ -1206,6 +1206,18 @@ def test_sort(dummy: Tensor) -> Tensor:
...
+    return ep.topk(t, 2)[0]
```
Comment (jonasrauber) [inline tests/test_main.py:1212]: Maybe you can also use tuple unpacking here:
`values, _ = ep.topk(t, 2)` and `return values`
(Same for the other test)

**Review by jonasrauber — CHANGES_REQUESTED** [review]
Could you make the stylistic changes I commented on (and add the better test)? Besides that, it looks good to me 👍

File: tests/test_main.py (hunk 4 above)
Comment (mglisse) [reply tests/test_main.py]: I had copied from the tests for `sort` and `argsort` where one is already sorted and the other reverse sorted, so I tested values on sorted and indices on reverse sorted. But I've changed it to something more complicated.

**Comment by mglisse** [comment]
Hello, are there still changes needed?

**Review by jonasrauber — APPROVED** [review]

**Comment by jonasrauber** [comment]
@mglisse Thanks, great work. Merged.
